File: Linked List/0430. Flatten a Multilevel Doubly Linked List/0430.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return

        stack = []
        dummy = Node(0,None,head,None)
        stack.append(head)
        prev = dummy

        while stack:
            node = stack.pop()

            node.prev = prev
            prev.next = node

            if node.next:
                stack.append(node.next)
                node.next = None
            if node.child:
                stack.append(node.child)
                node.child = None
            
            prev = node

        dummy.next.prev = None
        return dummy.next

File: Linked List/0430. Flatten a Multilevel Doubly Linked List/0430.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        if (!head) {
            return NULL;
        }
        Node* dummy = new Node(0,NULL,head,NULL);
        stack<Node*> nodeStack;
        nodeStack.push(head);
        Node* prev = dummy;

        while (!nodeStack.empty()) {
            Node* node = nodeStack.top();
            nodeStack.pop();
            node->prev = prev;
            prev->next = node;

            if (node->next) {
                nodeStack.push(node->next);
                node->next = NULL;
            } if (node->child) {
                nodeStack.push(node->child);
                node->child = NULL;
            }

            prev = node;
        }
        dummy->next->prev = NULL;
        return dummy->next;
        }
    };